File: extensions/shared/src/main/java/app/morphe/extension/youtube/patches/misc/OpenLinksDirectlyPatch.java
package app.morphe.extension.youtube.patches.misc;

import android.net.Uri;

import java.util.Objects;

import app.morphe.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public class OpenLinksDirectlyPatch {
    private static final String YOUTUBE_REDIRECT_PATH = "/redirect";

    public static Uri parseRedirectUri(String uri) {
        final Uri parsed = Uri.parse(uri);
        if (!Settings.BYPASS_URL_REDIRECTS.get())
            return parsed;

        if (Objects.equals(parsed.getPath(), YOUTUBE_REDIRECT_PATH)) {
            return Uri.parse(Uri.decode(parsed.getQueryParameter("q")));
        }

        return parsed;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/misc/openlink/OpenLinksDirectlyPatch.kt
package app.morphe.patches.reddit.misc.openlink

import app.morphe.patches.reddit.utils.extension.Constants
import app.morphe.patches.reddit.utils.patch.PatchList
import app.morphe.patches.reddit.utils.settings.settingsPatch
import app.morphe.patches.reddit.utils.settings.updatePatchStatus
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.patch.bytecodePatch

private const val EXTENSION_METHOD_DESCRIPTOR =
    "${Constants.PATCHES_PATH}/OpenLinksDirectlyPatch;" +
            "->" +
            "parseRedirectUri(Landroid/net/Uri;)Landroid/net/Uri;"

@Suppress("unused")
val openLinksDirectlyPatch = bytecodePatch(
    PatchList.OPEN_LINKS_DIRECTLY.title,
    PatchList.OPEN_LINKS_DIRECTLY.summary,
) {
    compatibleWith(app.morphe.patches.reddit.utils.compatibility.Constants.COMPATIBLE_PACKAGE)

    dependsOn(
        settingsPatch,
        screenNavigatorMethodResolverPatch
    )

    execute {
        screenNavigatorMethod.addInstructions(
            0, """
                invoke-static {p2}, $EXTENSION_METHOD_DESCRIPTOR
                move-result-object p2
                """
        )

        updatePatchStatus(
            "enableOpenLinksDirectly",
            PatchList.OPEN_LINKS_DIRECTLY
        )
    }
}
